### PR TITLE
fixing git clone retry mechanism to allow retries.

### PR DESCRIPTION
**Pull request by kobymeir**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-7901

## Description
fixing git clone retry mechanism to allow retries.

### CODE REVIEW

File: Tests/scripts/download_conf_repos.sh [hunk 1]
```diff
@@ -1,11 +1,53 @@
...
+  local branch=$2
+  local retry_count=$3
+  local exit_code=0
+  local i=1
+  echo "clone ${repo_name} from branch: ${branch} with ${retry_count} retries"
+  for ((i=1; i <= retry_count; i++)); do
+    git clone --depth=1 "https://gitlab-ci-token:${CI_JOB_TOKEN}@${repo_name}.git" --branch "${branch}" && exit_code=0 && break || exit_code=$?
+    if [ ${i} -ne "${retry_count}" ]; then
+      echo "Failed to clone ${repo_name} with branch:${branch}, exit code:${exit_code}, sleeping for 10 seconds and trying again"
+      sleep 10
```
Comment (tkatzir) [inline Tests/scripts/download_conf_repos.sh:15]: Does it make sense to keep the `10` in a local variable, like `i`?
So it's easy to change.

File: Tests/scripts/download_conf_repos.sh [hunk 2]
```diff
@@ -1,11 +1,53 @@
...
+    else
+      echo "Failed to clone ${repo_name} with branch:${branch}, exit code:${exit_code}, exhausted all retries"
```
Comment (tkatzir) [inline Tests/scripts/download_conf_repos.sh:17]:
```suggestion
      echo "Failed to clone ${repo_name} with branch:${branch}, exit code:${exit_code}, exhausted all ${retry_count} retries"
```

File: Tests/scripts/download_conf_repos.sh [hunk 3]
```diff
@@ -1,11 +1,53 @@
...
+      break
+    fi
+  done
+  if [ ${exit_code} -ne 0 ]; then
+    branch="master"
+    echo "Failed to clone ${repo_name} with branch:${branch}, will try to clone with branch:${branch}"
```
Comment (tkatzir) [inline Tests/scripts/download_conf_repos.sh:23]: Typo? This `branch` overrides the `branch` from above, and is used twice in the `echo`.
Also further down.
```suggestion
    master_branch="master"
    echo "Failed to clone ${repo_name} with branch ${branch}, will try to clone with branch ${master_branch}"
```

File: Tests/scripts/download_conf_repos.sh [hunk 4]
```diff
@@ -1,11 +1,53 @@
...
+    for ((i=1; i <= retry_count; i++)); do
```
Comment (tkatzir) [inline Tests/scripts/download_conf_repos.sh:24]: This block seems a Copy&Paste of the above.
Can they be extracted to another method?

**Review by tkatzir — COMMENTED** [review]
Looks good, not approved yet, some important comments.

File: Tests/scripts/download_conf_repos.sh [hunk 5]
```diff
@@ -1,11 +1,57 @@
...
+  local repo_name=$1
+  local branch=$2
+  local fallback_branch="${3:-master}"
+  local retry_count=$4
+  local sleep_time=${5:-10}  # default sleep time is 10 seconds.
+
+  clone_repository "${repo_name}" "${branch}" "${retry_count}" "${sleep_time}"
+  local exit_code=$?
+  if [ ${exit_code} -ne 0 ]; then
+    # Failed to clone with branch, try again with fallback_branch.
```
Comment (tkatzir) [inline Tests/scripts/download_conf_repos.sh:35]:
```suggestion
    # Failed to clone with branch, try again with fallback_branch.
    echo "Failed to clone ${repo_name} with branch ${branch}, exit code:${exit_code}, trying to clone with branch ${fallback_branch}!"
```

**Review by tkatzir — APPROVED** [review]
Very good! :-)
Approved, with one minor comment.

**Comment by kobymeir** [comment]
The build is failing on unrelated install issues.

**Pull request merged**